issue with select_all usage in regression methods

```python
mt = hl.utils.range_matrix_table(1, 1).annotate_cols(y=0, cov1=1).annotate_entries(x=1)

# x_expr uses a col field
x_expr = hl.case().when(mt.cov1 == 1, 2).default(mt.x)

hl.logistic_regression_rows(y=mt.y, x=x_expr, test='wald', covariates=[1])
```

```
Java stack trace:
java.lang.AssertionError: assertion failed: cov1 not in struct{__y: float64, __cov0: float64}
	at scala.Predef$.assert(Predef.scala:170)
	at is.hail.expr.ir.Infer$.apply(Infer.scala:74)
	at is.hail.expr.ir.InferIR$class.typ(IR.scala:62)
	at is.hail.expr.ir.GetField.typ(IR.scala:215)
	at is.hail.expr.ir.IRParser$.ir_value_expr_1(Parser.scala:561)
	at is.hail.expr.ir.IRParser$.ir_value_expr(Parser.scala:511)
	at is.hail.expr.ir.IRParser$.ir_value_expr_1(Parser.scala:536)
	at is.hail.expr.ir.IRParser$.ir_value_expr(Parser.scala:511)
	at is.hail.expr.ir.IRParser$$anonfun$ir_value_children$1.apply(Parser.scala:507)
	at is.hail.expr.ir.IRParser$$anonfun$ir_value_children$1.apply(Parser.scala:507)
	at is.hail.expr.ir.IRParser$.repUntil(Parser.scala:281)
	at is.hail.expr.ir.IRParser$.ir_value_children(Parser.scala:507)
	at is.hail.expr.ir.IRParser$.ir_value_expr_1(Parser.scala:714)
	at is.hail.expr.ir.IRParser$.ir_value_expr(Parser.scala:511)
	at is.hail.expr.ir.IRParser$.named_value_ir(Parser.scala:494)
	at is.hail.expr.ir.IRParser$$anonfun$named_value_irs$1.apply(Parser.scala:489)
	at is.hail.expr.ir.IRParser$$anonfun$named_value_irs$1.apply(Parser.scala:489)
	at is.hail.expr.ir.IRParser$.repUntil(Parser.scala:281)
	at is.hail.expr.ir.IRParser$.named_value_irs(Parser.scala:489)
	at is.hail.expr.ir.IRParser$.ir_value_expr_1(Parser.scala:670)
	at is.hail.expr.ir.IRParser$.ir_value_expr(Parser.scala:511)
	at is.hail.expr.ir.IRParser$.matrix_ir_1(Parser.scala:920)
	at is.hail.expr.ir.IRParser$.matrix_ir(Parser.scala:885)
	at is.hail.expr.ir.IRParser$.matrix_ir_1(Parser.scala:923)
	at is.hail.expr.ir.IRParser$.matrix_ir(Parser.scala:885)
	at is.hail.expr.ir.IRParser$$anonfun$parse_matrix_ir$2.apply(Parser.scala:1023)
	at is.hail.expr.ir.IRParser$$anonfun$parse_matrix_ir$2.apply(Parser.scala:1023)
	at is.hail.expr.ir.IRParser$.parse(Parser.scala:1007)
	at is.hail.expr.ir.IRParser$.parse_matrix_ir(Parser.scala:1023)
	at is.hail.expr.ir.IRParser$.parse_matrix_ir(Parser.scala:1022)
	at is.hail.expr.ir.IRParser.parse_matrix_ir(Parser.scala)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at py4j.reflection.MethodInvoker.invoke(MethodInvoker.java:244)
	at py4j.reflection.ReflectionEngine.invoke(ReflectionEngine.java:357)
	at py4j.Gateway.invoke(Gateway.java:280)
	at py4j.commands.AbstractCommand.invokeMethod(AbstractCommand.java:132)
	at py4j.commands.CallCommand.execute(CallCommand.java:79)
	at py4j.GatewayConnection.run(GatewayConnection.java:214)
	at java.lang.Thread.run(Thread.java:748)



Hail version: 0.2.5-e017bef1c26c
Error summary: AssertionError: assertion failed: cov1 not in struct{__y: float64, __cov0: float64}
```

--- a/hail/python/hail/matrixtable.py
+++ b/hail/python/hail/matrixtable.py
@@ -2580,34 +2580,29 @@ class MatrixTable(ExprContainer):
                     entry_exprs={},
                     global_exprs={},
                     ) -> 'MatrixTable':
-        all_exprs = list(itertools.chain(row_exprs.values(),
-                                         col_exprs.values(),
-                                         entry_exprs.values(),
-                                         global_exprs.values()))
 
-        mt = self._annotate_all(row_exprs,
-                                col_exprs,
-                                entry_exprs,
-                                global_exprs)
+        all_names = list(itertools.chain(row_exprs.keys(),
+                                         col_exprs.keys(),
+                                         entry_exprs.keys(),
+                                         global_exprs.keys()))
+        uids = {k: Env.get_uid() for k in all_names}
+
+        mt = self._annotate_all({uids[k]: v for k, v in row_exprs.items()},
+                                {uids[k]: v for k, v in col_exprs.items()},
+                                {uids[k]: v for k, v in entry_exprs.items()},
+                                {uids[k]: v for k, v in global_exprs.items()})
 
         if row_key is None:
             row_key = []
-        mt = mt.key_rows_by(*row_key)
+        mt = mt.key_rows_by(*(uids[k] for k in row_key))
 
         if col_key is None:
             col_key = []
-        mt = mt.key_cols_by(*col_key)
+        mt = mt.key_cols_by(*(uids[k] for k in col_key))
 
-        fields_to_keep = set(
-            list(mt.row_key)
-            + list(mt.col_key)
-            + list(row_exprs)
-            + list(col_exprs)
-            + list(entry_exprs)
-            + list(global_exprs)
-        )
-
-        return mt.drop(*(f for f in mt._fields if f not in fields_to_keep))
+        uid_set = set(uids.values())
+        return (mt.drop(*(f for f in mt._fields if f not in uid_set)) \
+                .rename({uid: original for original, uid in uids.items()}))
 
     def _process_joins(self, *exprs):
         return process_joins(self, exprs)

--- a/hail/python/hail/matrixtable.py
+++ b/hail/python/hail/matrixtable.py
@@ -2541,6 +2541,13 @@ class MatrixTable(ExprContainer):
                                          entry_exprs.values(),
                                          global_exprs.values()))
 
+        for field_name in list(itertools.chain(row_exprs.keys(),
+                                               col_exprs.keys(),
+                                               entry_exprs.keys(),
+                                               global_exprs.keys())):
+            if field_name in self._fields:
+                raise RuntimeError(f'field {repr(field_name)} already in matrix table, cannot use _annotate_all')
+
         base, cleanup = self._process_joins(*all_exprs)
         mir = base._mir
 
@@ -2578,30 +2585,29 @@ class MatrixTable(ExprContainer):
                                          entry_exprs.values(),
                                          global_exprs.values()))
 
-        base, cleanup = self._process_joins(*all_exprs)
-        mir = base._mir
+        mt = self._annotate_all(row_exprs,
+                                col_exprs,
+                                entry_exprs,
+                                global_exprs)
 
-        if row_key is not None:
-            mir = MatrixKeyRowsBy(mir, [])
-        row_struct = hl.struct(**row_exprs)
-        analyze("MatrixTable.select_rows", row_struct, self._row_indices)
-        mir = MatrixMapRows(mir, row_struct._ir)
-        if row_key is not None:
-            mir = MatrixKeyRowsBy(mir, row_key)
+        if row_key is None:
+            row_key = []
+        mt = mt.key_rows_by(*row_key)
 
-        col_struct = hl.struct(**col_exprs)
-        analyze("MatrixTable.select_cols", col_struct, self._col_indices)
-        mir = MatrixMapCols(mir, col_struct._ir, col_key)
+        if col_key is None:
+            col_key = []
+        mt = mt.key_cols_by(*col_key)
 
-        entry_struct = hl.struct(**entry_exprs)
-        analyze("MatrixTable.select_entries", entry_struct, self._entry_indices)
-        mir = MatrixMapEntries(mir, entry_struct._ir)
+        fields_to_keep = set(
+            list(mt.row_key)
+            + list(mt.col_key)
+            + list(row_exprs)
+            + list(col_exprs)
+            + list(entry_exprs)
+            + list(global_exprs)
+        )
 
-        globals_struct = hl.struct(**global_exprs)
-        analyze("MatrixTable.select_globals", globals_struct, self._global_indices)
-        mir = MatrixMapGlobals(mir, globals_struct._ir)
-
-        return cleanup(MatrixTable(mir))
+        return mt.drop(*(f for f in mt._fields if f not in fields_to_keep))
 
     def _process_joins(self, *exprs):
         return process_joins(self, exprs)

--- a/hail/python/hail/matrixtable.py
+++ b/hail/python/hail/matrixtable.py
@@ -2592,16 +2592,21 @@ class MatrixTable(ExprContainer):
                                 {uids[k]: v for k, v in entry_exprs.items()},
                                 {uids[k]: v for k, v in global_exprs.items()})
 
-        if row_key is None:
-            row_key = []
-        mt = mt.key_rows_by(*(uids[k] for k in row_key))
+        keep = set()
+        if row_key is not None:
+            old_key = list(mt.row_key)
+            mt = mt.key_rows_by(*(uids[k] for k in row_key)).drop(*old_key)
+        else:
+            keep = keep.union(set(mt.row_key))
 
-        if col_key is None:
-            col_key = []
-        mt = mt.key_cols_by(*(uids[k] for k in col_key))
+        if col_key is not None:
+            old_key = list(mt.col_key)
+            mt = mt.key_cols_by(*(uids[k] for k in col_key)).drop(*old_key)
+        else:
+            keep = keep.union(set(mt.col_key))
 
-        uid_set = set(uids.values())
-        return (mt.drop(*(f for f in mt._fields if f not in uid_set)) \
+        keep = keep.union(uids.values())
+        return (mt.drop(*(f for f in mt._fields if f not in keep)) \
                 .rename({uid: original for original, uid in uids.items()}))
 
     def _process_joins(self, *exprs):

--- a/hail/python/hail/methods/impex.py
+++ b/hail/python/hail/methods/impex.py
@@ -147,9 +147,7 @@ def export_gen(dataset, output, precision=4, gp=None, id1=None, id2=None,
     l = dataset.locus
     a = dataset.alleles
 
-    gen_exprs = {'locus': l,
-                 'alleles': a,
-                 'varid': expr_or_else(varid, hl.delimit([l.contig, hl.str(l.position), a[0], a[1]], ':')),
+    gen_exprs = {'varid': expr_or_else(varid, hl.delimit([l.contig, hl.str(l.position), a[0], a[1]], ':')),
                  'rsid': expr_or_else(rsid, ".")}
 
     for exprs, axis in [(sample_exprs, dataset._col_indices),
@@ -276,9 +274,7 @@ def export_plink(dataset, output, call=None, fam_id=None, ind_id=None, pat_id=No
     l = dataset.locus
     a = dataset.alleles
 
-    bim_exprs = {'locus': l,
-                 'alleles': a,
-                 'varid': expr_or_else(varid, hl.delimit([l.contig, hl.str(l.position), a[0], a[1]], ':')),
+    bim_exprs = {'varid': expr_or_else(varid, hl.delimit([l.contig, hl.str(l.position), a[0], a[1]], ':')),
                  'cm_position': expr_or_else(cm_position, 0.0)}
 
     for exprs, axis in [(fam_exprs, dataset._col_indices),

--- a/hail/python/hail/methods/statgen.py
+++ b/hail/python/hail/methods/statgen.py
@@ -422,7 +422,7 @@ def linear_regression_rows(y, x, covariates, block_size=16, pass_through=()) -> 
         x_field_name,
         cov_field_names,
         block_size,
-        list(row_fields)[len(mt.row_key):])
+        [x for x in row_fields if x not in mt.row_key])
 
     ht_result = Table._from_java(jt)
 
@@ -677,7 +677,7 @@ def logistic_regression_rows(test, y, x, covariates, pass_through=()) -> hail.Ta
         y_field_name,
         x_field_name,
         cov_field_names,
-        list(row_fields)[len(mt.row_key):])
+        [x for x in row_fields if x not in mt.row_key])
     return Table._from_java(jt)
 
 
@@ -750,7 +750,7 @@ def poisson_regression_rows(test, y, x, covariates, pass_through=()) -> Table:
         y_field_name,
         x_field_name,
         cov_field_names,
-        list(row_fields)[len(mt.row_key):])
+        [x for x in row_fields if x not in mt.row_key])
     return Table._from_java(jt)
 
 

--- a/hail/python/hail/methods/statgen.py
+++ b/hail/python/hail/methods/statgen.py
@@ -236,7 +236,7 @@ def impute_sex(call, aaf_threshold=0.0, include_par=False, female_threshold=0.2,
 
 def _get_regression_row_fields(mt, pass_through, method) -> Dict[str, str]:
 
-    row_fields = {}
+    row_fields = dict(zip(mt.row_key.keys(), mt.row_key.keys()))
     for f in pass_through:
         if isinstance(f, str):
             if f not in mt.row:
@@ -260,6 +260,8 @@ def _get_regression_row_fields(mt, pass_through, method) -> Dict[str, str]:
                 if not (name in mt.row_key and f._ir == mt[name]._ir):
                     raise ValueError(f"'{method}/pass_through': found duplicated field {repr(name)}")
             row_fields[name] = f
+    for k in mt.row_key:
+        del row_fields[k]
     return row_fields
 
 

--- a/hail/python/hail/methods/statgen.py
+++ b/hail/python/hail/methods/statgen.py
@@ -2121,6 +2121,7 @@ def split_multi_hts(ds, keep_star=False, left_aligned=False, vep_root='vep'):
     if isinstance(ds, Table):
         return split.annotate(**update_rows_expression).drop('old_locus', 'old_alleles')
 
+    split = split.annotate_rows(**update_rows_expression)
     entry_fields = ds.entry
 
     expected_field_types = {
@@ -2169,9 +2170,7 @@ def split_multi_hts(ds, keep_star=False, left_aligned=False, vep_root='vep'):
         update_entries_expression['PGT'] = hl.downcode(split.PGT, split.a_index)
     if 'PID' in entry_fields:
         update_entries_expression['PID'] = split.PID
-    return split._annotate_all(
-        row_exprs=update_rows_expression,
-        entry_exprs=update_entries_expression).drop('old_locus', 'old_alleles')
+    return split.annotate_entries(**update_entries_expression).drop('old_locus', 'old_alleles')
 
 
 @typecheck(call_expr=expr_call)

--- a/hail/python/hail/methods/statgen.py
+++ b/hail/python/hail/methods/statgen.py
@@ -236,8 +236,7 @@ def impute_sex(call, aaf_threshold=0.0, include_par=False, female_threshold=0.2,
 
 def _get_regression_row_fields(mt, pass_through, method) -> Dict[str, str]:
 
-    # include key as base
-    row_fields = dict(mt.row_key)
+    row_fields = {}
     for f in pass_through:
         if isinstance(f, str):
             if f not in mt.row:
@@ -1090,8 +1089,6 @@ def linear_mixed_regression_rows(entry_expr,
                                 a_t_path if model.low_rank else None,
                                 partition_size)
     row_fields = _get_regression_row_fields(mt, pass_through, 'linear_mixed_regression_rows')
-    for k in mt.row_key:
-        del row_fields[k]
 
     mt_keys = mt.select_rows(**row_fields).add_row_index('__row_idx').rows().add_index('__row_idx').key_by('__row_idx')
     return mt_keys.annotate(**ht[mt_keys['__row_idx']]).key_by(*mt.row_key).drop('__row_idx')

--- a/hail/python/test/hail/methods/test_impex.py
+++ b/hail/python/test/hail/methods/test_impex.py
@@ -930,7 +930,7 @@ class GENTests(unittest.TestCase):
         in1 = (hl.import_gen(out1 + '.gen', sample_file=out1 + '.sample', min_partitions=3)
                .add_col_index()
                .add_row_index())
-        self.assertTrue(in1.aggregate_entries(hl.agg.fraction(in1.GP == [0.0, 1.0, 0.0])) == 1.0)
+        self.assertTrue(in1.aggregate_entries(hl.agg.fraction((hl.is_missing(in1.GP) | (in1.GP == [0.0, 1.0, 0.0])) == 1.0)))
         self.assertTrue(in1.aggregate_rows(hl.agg.fraction((in1.varid == hl.str(in1.row_idx)) &
                                                            (in1.rsid == hl.str(in1.row_idx)))) == 1.0)
         self.assertTrue(in1.aggregate_cols(hl.agg.fraction((in1.s == hl.str(in1.col_idx)))))

--- a/hail/python/test/hail/methods/test_statgen.py
+++ b/hail/python/test/hail/methods/test_statgen.py
@@ -1465,3 +1465,14 @@ class Tests(unittest.TestCase):
                            [mt.row_idx, mt.col_idx, mt.GT.n_alt_alleles()]]:
             self.assertTrue(hl.methods.statgen._warn_if_no_intercept('', covariates))
             self.assertFalse(hl.methods.statgen._warn_if_no_intercept('', [intercept] + covariates))
+
+    def test_regression_field_dependence(self):
+        mt = hl.utils.range_matrix_table(10, 10)
+        mt = mt.annotate_cols(c1 = hl.literal([x % 2 == 0 for x in range(10)])[mt.col_idx], c2 = hl.rand_norm(0, 1))
+        mt = mt.annotate_entries(e1 = hl.int(hl.rand_norm(0, 1) * 10))
+
+        x_expr = hl.case().when(mt.c2 < 0, 0).default(mt.e1)
+
+        hl.logistic_regression_rows('wald', y=mt.c1, x=x_expr, covariates=[1])
+        hl.poisson_regression_rows('wald', y=mt.c1, x=x_expr, covariates=[1])
+        hl.linear_regression_rows(y=mt.c1, x=x_expr, covariates=[1])

